File: ruby/lib/coding_interview/infrastructure/repository/account_repository_impl.rb
require_relative "../../application/repository/account_repository"

module CodingInterview
  module Infrastructure
    module Repository
      class AccountRepositoryImpl < Application::Repository::AccountRepository
        def initialize
          @store = {}
          @mutex = Mutex.new
        end

        def find(user_id)
          @mutex.synchronize { @store[user_id.value] }
        end

        def upsert(user_id, account)
          @mutex.synchronize { @store[user_id.value] = account }
          nil
        end

        def exists?(user_id)
          @mutex.synchronize { @store.key?(user_id.value) }
        end
      end
    end
  end
end

File: python/src/coding_interview/infrastructure/repository/account_repository_impl.py
from __future__ import annotations

from coding_interview.domain.stock import Account
from coding_interview.domain.user_id import UserId


class AccountRepositoryImpl:
    def __init__(self) -> None:
        self._store: dict[str, Account] = {}

    def find(self, user_id: UserId) -> Account | None:
        return self._store.get(user_id.value)

    def upsert(self, user_id: UserId, account: Account) -> None:
        self._store[user_id.value] = account

    def exists(self, user_id: UserId) -> bool:
        return user_id.value in self._store
